Frameworks/Foundation/NSScanner.mm

```diff
@@ -575,147 +339,249 @@ - (BOOL)scanUpToString:(NSString*)string intoString:(NSString**)intoString {
-
-        if ([_skipSet characterIsMember:unicode] && scanStarted == NO)
-            continue;
-        else if ([charset characterIsMember:unicode])
-            break;
-        else {
-            scanStarted = YES;
-            result[resultLength++] = unicode;
-        }
-    }
-
-    if (resultLength > 0) {
-        if (stringp) {
-            *stringp = [NSString stringWithCharacters:result length:resultLength];
-        }
-
-        IwFree(result);
-        return YES;
-    } else {
-        _location = oldLocation;
-
-        IwFree(result);
-        return NO;
-    }
+- (BOOL)scanUpToCharactersFromSet:(NSCharacterSet* _Nonnull)charset intoString:(NSString* _Nullable*)stringp {
+    return [self scanCharactersFromSet:[charset invertedSet] intoString:stringp];
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (BOOL)scanDecimal:(NSDecimal*)decimalValue {
-    UNIMPLEMENTED();
-    return StubReturn();
-}
+    // TODO #2369: This could probably be made more precise if it scanned directly from string -> NSDecimalNumber
+    // but WinObjC does not have [NSDecimalNumber decimalNumberWithString:] yet
+    double doubleValue;
+    if (![self scanDouble:&doubleValue]) {
+        return NO;
+    }
 
-/**
- @Status Stub
- @Notes
-*/
-- (BOOL)scanHexDouble:(double*)result {
-    UNIMPLEMENTED();
-    return StubReturn();
-}
+    if (decimalValue) {
+        *decimalValue = [[NSDecimalNumber numberWithDouble:doubleValue] decimalValue];
+    }
 
-/**
- @Status Stub
- @Notes
-*/
-- (BOOL)scanHexFloat:(float*)result {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return YES;
 }
 
 /**
  @Status Interoperable
  @Notes
 */
-- (BOOL)scanUnsignedLongLong:(unsigned long long*)pValue {
-    const char* pScanStart = (char*)[_string UTF8String];
-    char* pScanEnd = nullptr;
+- (BOOL)scanHexDouble:(double*)result {
+    enum {
+        STATE_START,
+        STATE_SIGN,
+        STATE_ZERO,
+        STATE_MANTISSA,
+        STATE_P,
+        STATE_EXPONENT,
+    } state = STATE_START;
+
+    bool mantissaIsNegative = false;
+    bool exponentIsNegative = false;
+    bool scannedPeriod = false;
+    bool hasValue = false;
+
+    double mantissa = 0;
+    int exponent = 0;
+
+    // Once a '.' has been scanned, this /= 16 every character, so that each base-16 decimal digit is appropriately placed
+    // ie: 1.1ae = 1 + (1 * (mantissaDigitPlace /= 16)) + (a * (mantissaDigitPlace /= 16)) + (e * (mantissaDigitPlace /= 16))
+    double mantissaDigitPlace = 1;
+
+    const NSUInteger length = _string.length;
+    NSUInteger i = [self _indexOfNextUnskippedCharacter];
+    for (; i < length; ++i) {
+        unichar unicode = _stringChars[i];
+
+        switch (state) {
+            case STATE_START:
+                if (unicode == '-') {
+                    state = STATE_SIGN;
+                    mantissaIsNegative = true;
+                } else if (unicode == '+') {
+                    state = STATE_SIGN;
```

Comment:
hmm maybe not, given this is being looped over.

---
In reply to: [113068748](https://github.com/Microsoft/WinObjC/pull/2582#discussion_r113068748) [](ancestors = 113068748)